catalogue/webapp/services/prismic/index.ts

```diff
@@ -0,0 +1,153 @@
+
+export type PrismicQueryProps = {
+  query?: string | string[]
+  pageSize?: number;
+};
+
+export const prismicApiError = (): PrismicApiError => ({
+  errorType: 'http',
+  httpStatus: 500,
+  label: 'Internal Server Error',
+  description: '',
+  type: 'Error',
+});
+
+const agentKeepAlive = new Agent({
+  keepAlive: true,
+  freeSocketTimeout: 1000 * 59, // 1s less than the akka-http idle timeout
+});
+
+// TODO: move Prismic fetch function out to common and use it here and in server-data/prismic
+// To authenticate a request to the Prismic API graphql endpoint you need a Prismic-ref value that refreshes every 30 seconds.
+// The ref can be used in a 'stale' state for 5 minutes before it expires
+// https://community.prismic.io/t/for-how-long-is-it-safe-to-cache-a-prismic-api-ref/5962
+export const prismicRefFetch = (
+  url: string,
+  options?: Record<string, string>
+): Promise<Response> => {
+  return fetch(url, { ...options, agent: agentKeepAlive });
+};
+
+// Prismic API graphql endpoint expects query requests via GET method
+// This means we need to be able to pass a body in the request, something that is not possible with fetch
+// As a workaround I have used axios to make the request
+export const prismicFetch = (
+  options?: AxiosRequestConfig
+): Promise<AxiosResponse> => {
+  return axios({ ...options });
+};
+
+export async function prismicQuery(
+  endpoint: string,
+  { query, pageSize }: PrismicQueryProps
+): Promise<StoryResultsList<Story> | PrismicApiError> {
+  const graphQuery = `query {
+  allArticless(fulltext: "${query}" sortBy: title_ASC first: ${pageSize}) {
+  edges {
+      node {
+        title
+        _meta { id }
+        contributors {
+          contributor {
+            ...on People {
+              name
+            }
+          }
+        }
+        body {
+          ...on ArticlesBodyStandfirst {
+            primary {
+              text
+            }
+          }
+        }
+        promo {
+          ...on ArticlesPromoEditorialimage {
+            primary {
+              image
+              link
+              caption
+            }
+          }
+        }
+      }
+    }
+  }
+}`;
+  const url = `https://wellcomecollection.prismic.io/graphql?query=${graphQuery}`;
+  const fetchRefUrl = 'https://wellcomecollection.prismic.io/api/v2';
+
+  const headers = {
+    Authorization: `Bearer ${process.env.PRISMIC_BEARER_TOKEN}`,
+    repository: 'wellcomecollection',
+  };
+
+  const prismicRef = await prismicRefFetch(fetchRefUrl, headers);
+  const { refs } = await prismicRef.json();
+  const { ref } = refs[0];
+  headers['Prismic-ref'] = ref;
+  const updatedHeaders = {
+    Authorization: `Bearer ${process.env.PRISMIC_TOKEN}`,
+    repository: 'wellcomecollection',
+    'Prismic-ref': ref,
+  };
+  const options: AxiosRequestConfig = {
+    method: 'GET',
+    url: url,
+    headers: updatedHeaders,
+    data: {
+      query: { graphQuery },
+    },
+  };
```

Comment:
This (and the rest of the file, to a lesser extent) seems to replicate much of the functionality of `@prismicio/client`, which is what we use elsewhere for Prismic - is there missing functionality in that client that we need to implement ourselves?

Would it be possible to reuse our existing client (therefore getting all the ref/caching goodness) and use the [`graphQLFetch`](https://github.com/prismicio/prismic-client/blob/master/src/client.ts#L1448) method? Thus avoiding having to implement this stuff ourselves, worry about axios vs fetch, etc etc...

Reply:
> This (and the rest of the file, to a lesser extent) seems to replicate much of the functionality of `@prismicio/client`, which is what we use elsewhere for Prismic - is there missing functionality in that client that we need to implement ourselves?
> 
> Would it be possible to reuse our existing client (therefore getting all the ref/caching goodness) and use the [`graphQLFetch`](https://github.com/prismicio/prismic-client/blob/master/src/client.ts#L1448) method? Thus avoiding having to implement this stuff ourselves, worry about axios vs fetch, etc etc...

Thanks for the steer on this. I have massively reduced the complexity of recreating what `@prismicio/client` does for us. I dug through the reasoning behind why the engineers added the graphql methods [here](https://github.com/prismicio/prismic-client/issues/202#issue-1077360035). The issue suggests that integrating a graphql client with the `@prismicio/client` is the way to go, so I have opted for [`graphql-request`](https://yarnpkg.com/package/graphql-request).